### PR TITLE
IGNITE-12383 Add distances measures between two vectors

### PR DESCRIPTION
**Pull request by dehasi**
#### The following measures were added:

* [Chebyshev distance](http://en.wikipedia.org/wiki/Chebyshev_distance)
* [Cosine similarity](http://en.wikipedia.org/wiki/Cosine_similarity)
* [Minkowski distance](http://en.wikipedia.org/wiki/Minkowski_distance)
* [Jaccard index](http://en.wikipedia.org/wiki/Jaccard_index)

#### Testing
Property tests for distance measures (see unit-tests)  is added.

#### Refactoring
The default implementation for a couple of methods (see `DistanceMeasure`) is added as well.

#### To discuss
As Jaccard index and cosine similarity are not actually a distance measure (not all properties of a distance measure are applicable for them) should they implement `DistanceMeasure` interface?

#### Question
To implement the Jaccard index I need to calculate intersection. I used `HashSet`. Is there a more efficient way to compute intersection of two vectors?

@zaleslaw @avplatonov could you take a look?

### CODE REVIEW

File: modules/ml/src/main/java/org/apache/ignite/ml/math/distances/CosineSimilarity.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * Calculates the {@code A * B / (|A| * |B|)} (Cosine similarity) distance between two points.
+ */
+public class CosineSimilarity implements DistanceMeasure {
+
+    /** {@inheritDoc} */
+    @Override
+    public double compute(Vector a, Vector b) throws CardinalityException {
+        return MatrixUtil.localCopyOf(a).dot(b) / (a.kNorm(2d) * b.kNorm(2d));
+    }
+
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/math/distances/CosineSimilarity.java]: remove this line

File: modules/ml/src/main/java/org/apache/ignite/ml/math/distances/ChebyshevDistance.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+import org.apache.ignite.ml.math.exceptions.math.CardinalityException;
+import org.apache.ignite.ml.math.primitives.vector.Vector;
+import org.apache.ignite.ml.math.util.MatrixUtil;
+
+/**
+ * Calculates the {@code max(x_i - y_i)} (Chebyshev) distance between two points.
+ */
+public class ChebyshevDistance implements DistanceMeasure {
+
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/math/distances/ChebyshevDistance.java]: remove this line

File: modules/ml/src/main/java/org/apache/ignite/ml/math/distances/CosineSimilarity.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import org.apache.ignite.ml.math.exceptions.math.CardinalityException;
+import org.apache.ignite.ml.math.primitives.vector.Vector;
+import org.apache.ignite.ml.math.util.MatrixUtil;
+
+/**
+ * Calculates the {@code A * B / (|A| * |B|)} (Cosine similarity) distance between two points.
+ */
+public class CosineSimilarity implements DistanceMeasure {
+
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/math/distances/CosineSimilarity.java]: remove this line

File: modules/ml/src/main/java/org/apache/ignite/ml/math/distances/JaccardIndex.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.HashSet;
+import java.util.Set;
+import org.apache.ignite.ml.math.exceptions.math.CardinalityException;
+import org.apache.ignite.ml.math.primitives.vector.Vector;
+
+/**
+ * Calculates {@code J = |A \cap B| / |A \cup B| } (Jaccard index) distance between two points.
+ */
+public class JaccardIndex implements DistanceMeasure {
+
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/math/distances/JaccardIndex.java]: remove this line

File: modules/ml/src/main/java/org/apache/ignite/ml/math/distances/JaccardIndex.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    @Override
+    public double compute(Vector a, Vector b) throws CardinalityException {
+        Set<Double> aSet = new HashSet<>();
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/math/distances/JaccardIndex.java]: rename the variable to something meaningful

File: modules/ml/src/test/java/org/apache/ignite/ml/math/distances/JaccardIndexTest.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+package org.apache.ignite.ml.math.distances;
+
+import org.apache.ignite.ml.math.primitives.vector.Vector;
+import org.apache.ignite.ml.math.primitives.vector.impl.DenseVector;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+/** Test for {@code JaccardIndex}. */
+public class JaccardIndexTest {
```
Comment (zaleslaw) [inline modules/ml/src/test/java/org/apache/ignite/ml/math/distances/JaccardIndexTest.java]: Did you add these tests to the common suite?

**Comment by zaleslaw** [comment]
I'll need to think about your questions @dehasi  it will be a ticket for 2.9 and we have enough time to decide what should be included or not. Thanks for the PR, please fix minor comments, I'll return in a month to this PR

File: modules/ml/src/test/java/org/apache/ignite/ml/math/distances/JaccardIndexTest.java (hunk 6 above)
Comment (dehasi) [reply modules/ml/src/test/java/org/apache/ignite/ml/math/distances/JaccardIndexTest.java]: I added both `JaccardIndexTest` and `CosineSimilarityTest` to `MathImplLocalTestSuite`. Thank you for reminding.

File: modules/ml/src/main/java/org/apache/ignite/ml/math/distances/ChebyshevDistance.java (hunk 2 above)
Comment (dehasi) [reply modules/ml/src/main/java/org/apache/ignite/ml/math/distances/ChebyshevDistance.java]: removed

File: modules/ml/src/main/java/org/apache/ignite/ml/math/distances/JaccardIndex.java (hunk 4 above)
Comment (dehasi) [reply modules/ml/src/main/java/org/apache/ignite/ml/math/distances/JaccardIndex.java]: removed.

File: modules/ml/src/main/java/org/apache/ignite/ml/math/distances/CosineSimilarity.java (hunk 3 above)
Comment (dehasi) [reply modules/ml/src/main/java/org/apache/ignite/ml/math/distances/CosineSimilarity.java]: Removed

File: modules/ml/src/main/java/org/apache/ignite/ml/math/distances/JaccardIndex.java (hunk 5 above)
Comment (dehasi) [reply modules/ml/src/main/java/org/apache/ignite/ml/math/distances/JaccardIndex.java]: It's hard :) I renamed `aSet` -> `uniqueValues`. If you have any better idea please, let me know.

**Comment by dehasi** [comment]
@zaleslaw Thank you for your review. I fixed minor comments. I don't mind if we postpone the discussion we have enough time.

**Comment by zaleslaw** [comment]
@dehasi try to revisit next week

**Comment by zaleslaw** [comment]
@dehasi PR is merged

**Pull request merged**